File: app/src/main/java/org/metabrainz/mobile/presentation/features/recording/RecordingInfoFragment.java
package org.metabrainz.mobile.presentation.features.recording;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import org.metabrainz.mobile.data.sources.api.entities.EntityUtils;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.MBEntity;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Recording;
import org.metabrainz.mobile.databinding.FragmentRecordingInfoBinding;

public class RecordingInfoFragment extends Fragment {

    private FragmentRecordingInfoBinding binding;
    private RecordingViewModel recordingViewModel;

    public static RecordingInfoFragment newInstance() {
        return new RecordingInfoFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentRecordingInfoBinding.inflate(inflater, container, false);
        recordingViewModel = new ViewModelProvider(requireActivity()).get(RecordingViewModel.class);
        recordingViewModel.getData().observe(getViewLifecycleOwner(), this::setRecordingInfo);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void setRecordingInfo(MBEntity entity) {
        if (entity instanceof Recording) {
            Recording recording = (Recording) entity;
            String duration, artist;
            binding.recordingTitle.setText(recording.getTitle());
            duration = recording.getDuration();
            artist = EntityUtils.getDisplayArtist(recording.getArtistCredits());
            if (duration != null) binding.recordingDuration.setText(duration);
            if (artist != null) binding.recordingArtist.setText(artist);
        }
    }
}

File: app/src/main/java/org/metabrainz/mobile/presentation/features/recording/RecordingInfoFragment.kt
package org.metabrainz.mobile.presentation.features.recording

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import org.metabrainz.mobile.data.sources.api.entities.EntityUtils
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Recording
import org.metabrainz.mobile.databinding.FragmentRecordingInfoBinding
import org.metabrainz.mobile.presentation.features.base.MusicBrainzFragment
import org.metabrainz.mobile.util.Resource

class RecordingInfoFragment : Fragment() {
    private var binding: FragmentRecordingInfoBinding? = null
    private val recordingViewModel: RecordingViewModel by activityViewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentRecordingInfoBinding.inflate(inflater, container, false)
        recordingViewModel.data.observe(viewLifecycleOwner) { setRecordingInfo(it) }
        return binding!!.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    private fun setRecordingInfo(resource: Resource<Recording>) {
        if (resource.status == Resource.Status.SUCCESS) {
            val recording = resource.data
            binding!!.recordingTitle.text = recording.title
            if (recording.duration != null) binding!!.recordingDuration.text = recording.duration
            if (recording.artistCredits != null)
                binding!!.recordingArtist.text = EntityUtils.getDisplayArtist(recording.artistCredits)
        }
    }

    companion object : MusicBrainzFragment {
        override fun newInstance(): Fragment {
            return RecordingInfoFragment()
        }
    }
}